File: runme/mcgen/local/datan/solsrc/s2tim.f
      PROGRAM S2TIM
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(MAXN=400,MAXL=10,MAXK=40)
      DIMENSION Y(MAXN),T(MAXN+2*MAXK),ETA(MAXN+2*MAXK)
      DIMENSION ETADEL(MAXN+2*MAXK),CONETA(MAXN+2*MAXK)
      DIMENSION ATA1(MAXL+1,MAXL+1),ATA1AT(MAXL+1,2*MAXK+1)
      DIMENSION SCRAT(MAXL+1,MAXL+1),A(2*MAXK+1,MAXL+1)
C COMMON /CS2TIM/ is needed for Microsoft FORTRAN compiler
      COMMON /CS2TIM/ Y,T,ETA,ETADEL,CONETA,ATA1,ATA1AT,SCRAT,A
      CHARACTER*1 STRING
      CHARACTER*75 CAPT
C identify program to user
      WRITE(*,*)' Program S2TIM simulates data with errors corres-'
      WRITE(*,*)' ponding to continuous or discontinuous functions,'
      WRITE(*,*)' performs time series analysis on them'
      WRITE(*,*)' and presents results graphically'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)'  Enter N(<= 400)'
      WRITE(*,*)' >'
      READ(*,*)N
      IF (N.GT.400) N=400
      WRITE(*,*)' Enter SIGMA (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMA
      WRITE(*,*)' Enter k ( 0 < k < min(40,N) )'
      WRITE(*,*)' >'
      READ(*,*)K
      IF (K.GT.40) K=40
      WRITE(*,*)' Enter l ( 0 < l < k  and l < 10)'
      WRITE(*,*)' >'
      READ(*,*)L
      IF (L.GT.MIN(K,9)) L=MIN(K,9)
      WRITE(*,*)' Enter P ( 0 < P < 1.0 )'
      WRITE(*,*)' >'
      READ(*,*)P
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - sine function'
      WRITE(*,*)' 2 - step function'
      WRITE(*,*)' 3 - saw tooth function'
      WRITE(*,*)' >'
      READ(*,*)ISWIT
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
C generate time series
      CALL RNSTNR(Y,N)
      DO 1 I=1,N
        Y(I)=Y(I)*SIGMA
        TIME=DBLE(I)
        IF(ISWIT.EQ.1) THEN
          Y(I)=SIN(.0174532*TIME)+Y(I)
        ELSE IF(ISWIT.EQ.2) THEN
          TT=MOD(TIME,200.D0)
          IF(TT.LE.100.D0) THEN
            Y(I)=-1.D0+Y(I)
          ELSE
            Y(I)=1.D0+Y(I)
          END IF
        ELSE IF(ISWIT.EQ.3) THEN
          TT=MOD(TIME,100.D0)
          Y(I)=0.02*(TT-50.D0)+Y(I)
        END IF
1     CONTINUE
      CALL TIMSER(Y,N,K,L,P,ETA,CONETA,A,ATA1,ATA1AT,SCRAT)
      CALL GROPEN
      CALL GROPWS(NWS)
      YMIN=Y(1)
      YMAX=Y(1)
      DO 7 I=2,N
        YMIN=MIN(Y(I),YMIN)
        YMAX=MAX(Y(I),YMAX)
7     CONTINUE
      CALL GRWNCC(-20.D0,DBLE(N+20),-1.5D0,1.5D0)
      CALL GRVWWC(0.D0,1.D0,.2D0,.8D0)
      CALL GRWNWC(-.314D0,1.1D0,0.D0,1.D0)
      CALL GRBOUN
      CALL GRFRAM
C write caption
      CAPT='N = ***, k = ***, l = ***, P = *****'
      WRITE(CAPT(5:7),'(I3)')N
      WRITE(CAPT(14:16),'(I3)')K
      WRITE(CAPT(23:25),'(I3)')L
      WRITE(CAPT(32:36),'(F5.3)')P
      CALL GRTXTC(1.D0,CAPT,75)
      STRING='t'
      CALL GRSCLX(STRING,1)
      STRING='y'
      CALL GRSCLY(STRING,1)
C draw data points
      CALL GRSTCL(2)
      DO 10 I=1,N
        CALL GRDATP(1,.2D0,DBLE(I),Y(I),0.D0,0.D0,0.D0)
10    CONTINUE
C draw moving averages
      CALL GRSTCL(3)
      DO 20 I=1,N+2*K
        T(I)=DBLE(I-K)
20    CONTINUE
      CALL GRPLIN(N+2*K,T,ETA)
      CALL GRSTCL(4)
C draw confidence limits
      DO 30 I=1,N+2*K
        ETADEL(I)=ETA(I)+CONETA(I)
30    CONTINUE
      CALL GRPLIN(N+2*K,T,ETADEL)
      DO 40 I=1,N+2*K
        ETADEL(I)=ETA(I)-CONETA(I)
40    CONTINUE
      CALL GRPLIN(N+2*K,T,ETADEL)
      CALL GRCLSE
      END

File: runme/mcgen/local/datan/solsrc/s2tim.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__3 = 3, c__4 = 4, c__75 = 75;
static double c_d200 = 200., c_d100 = 100., c_dm20 = -20., c_dm1p5 = -1.5,
              c_d1p5 = 1.5, c_d0 = 0., c_d1 = 1., c_dp2 = .2, c_dp8 = .8,
              c_dmp314 = -.314, c_d1p1 = 1.1;

main(void)
{
    /* Local variables */
    static integer i, k, l, n, iswit, nws;
#if !defined(__TURBOC__)
    static
#endif
    double p, sigma, time, ymin, ymax, tt,
                  y[400], t[480], eta[480], etadel[480], coneta[480],
                  ata1[121] /* was [11][11] */,
                  ata1at[891] /* was [11][81] */,
                  scrat[121] /* was [11][11] */,
                  a[891] /* was [81][11] */;
    static char capt[75], string[1];
    /* System generated locals */
    integer i__1;
    double d__1;
/* identify program to user */
    printf("%s\n", "Program S2TIM simulates data with errors corres-");
    printf("%s\n", "ponding to continuous or discontinuous functions,");
    printf("%s\n", "performs time series analysis on them");
    printf("%s\n\n", "and presents results graphically.");
/* ask for input */
    printf("%s\n", "Enter N(<= 400):");
    printf("%s", "> ");
    scanf("%li", &n);
    if (n > 400) n = 400;
    printf("%s\n", "Enter SIGMA (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigma);
    printf("%s\n", "Enter k ( 0 < k < min(40,N) ):");
    printf("%s", "> ");
    scanf("%li", &k);
    if (k > 40) k = 40;
    printf("%s\n", "Enter l ( 0 < l < k  and l < 10):");
    printf("%s", "> ");
    scanf("%li", &l);
    if (l > min(k,9)) l = min(k,9);
    printf("%s\n", "Enter P ( 0 < P < 1.0 ):");
    printf("%s", "> ");
    scanf("%lf", &p);
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - sine function");
    printf("%s\n", "2 - step function");
    printf("%s\n", "3 - saw tooth function");
    printf("%s", "> ");
    scanf("%li", &iswit);
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
/* generate time series */
    rnstnr_(y, &n);
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
        y[i - 1] *= sigma;
        time = (double) i;
        if (iswit == 1) {
            y[i - 1] = sin(time * .0174532f) + y[i - 1];
        } else if (iswit == 2) {
            tt = fmod(time, c_d200);
            if (tt <= 100.) {
                y[i - 1] += -1.;
            } else {
                y[i - 1] += 1.;
            }
        } else if (iswit == 3) {
            tt = fmod(time, c_d100);
            y[i - 1] = (tt - 50.) * .02f + y[i - 1];
        }
    }
    timser_(y, &n, &k, &l, &p, eta, coneta, a, ata1, ata1at, scrat);
    gropen_();
    gropws_(&nws);
    ymin = y[0];
    ymax = y[0];
    i__1 = n;
    for (i = 2; i <= i__1; ++i) {
/* Computing MIN */
        d__1 = y[i - 1];
        ymin = min(d__1,ymin);
/* Computing MAX */
        d__1 = y[i - 1];
        ymax = max(d__1,ymax);
    }
    d__1 = (double) (n + 20);
    grwncc_(&c_dm20, &d__1, &c_dm1p5, &c_d1p5);
    grvwwc_(&c_d0, &c_d1, &c_dp2, &c_dp8);
    grwnwc_(&c_dmp314, &c_d1p1, &c_d0, &c_d1);
    grboun_();
    grfram_();
/* write caption */
    sprintf(capt, "N = %3li, k = %3li, l = %3li, P = %5.3lf", n, k, l, p);
    memset(capt + 36, ' ', 39);
    grtxtc_(&c_d1, capt, &c__75);
    *(unsigned char *)string = 't';
    grsclx_(string, &c__1);
    *(unsigned char *)string = 'y';
    grscly_(string, &c__1);
/* draw data points */
    grstcl_(&c__2);
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
        d__1 = (double) i;
        grdatp_(&c__1, &c_dp2, &d__1, &y[i - 1], &c_d0, &c_d0, &c_d0);
    }
/* draw moving averages */
    grstcl_(&c__3);
    i__1 = n + (k << 1);
    for (i = 1; i <= i__1; ++i) {
        t[i - 1] = (double) (i - k);
    }
    i__1 = n + (k << 1);
    grplin_(&i__1, t, eta);
    grstcl_(&c__4);
/* draw confidence limits */
    i__1 = n + (k << 1);
    for (i = 1; i <= i__1; ++i) {
        etadel[i - 1] = eta[i - 1] + coneta[i - 1];
    }
    i__1 = n + (k << 1);
    grplin_(&i__1, t, etadel);
    i__1 = n + (k << 1);
    for (i = 1; i <= i__1; ++i) {
        etadel[i - 1] = eta[i - 1] - coneta[i - 1];
    }
    i__1 = n + (k << 1);
    grplin_(&i__1, t, etadel);
    grclse_();
    return 0;
} /* main */

